### ISSUE TITLE
Need help with replicating Figure 5 

### ISSUE THREAD

**Issue by schen572** [kampmichael/RelativeFlatnessAndGeneralization#1]
Good day, 

I was recently trying to replicate Figure 5 in the paper "relative flatness and generalization". I am not able to achieve error even below 0.9 even after extensive training (600 epochs). I noticed that when [loading ](https://github.com/kampmichael/RelativeFlatnessAndGeneralization/blob/0baa1f0c87db2860e1a4d8f675ff0347a8872b3f/CorrelationFlatnessGeneralization/data_loaders.py#L43)cifar10 data you did data augmentation, which keeps the training loss above 0.9. After commenting that out, I trained until the loss is below 0.5, and here is my result calculating using your code:
![image](https://github.com/kampmichael/RelativeFlatnessAndGeneralization/assets/43688484/8d212d55-b300-43d1-9fd9-ba829ca62989)

Now here are things that I need help with:
1. The difference between the test loss and train loss is below 1, why does the y axis of figure 5 range from 4-9? I cannot think of scaling of any sort that will throw the value into that range.
2. how many epochs did you train the LeNet5? What are the final test accuracy and train accuracy?
3. Figure 7 tells me that the magnitude of relative flatness should be around 1000, but the result I have is around 190. Could you please help me understand?

Again, thank you for your help.

**Comment by link-er** [comment]
Hello,

The data augmentation in CIFAR10 is necessary to achieve good generalization, without it the networks would be severely underfit.

1 - I am wondering how you got such small loss - the values we are using for plotting is elementwise loss, without mean reduction. So the overall training loss that is in your screenshot is way too small - did you edit some code? Which file do you run for these results?
2 - Up to my memory the training was taking much longer than 600 epochs. Since we selected a simplistic architecture (LeNet5) and complicated data (CIFAR10) one will need thousands of epochs to reach the small loss. The criterion was loss < 0.1. One of the examples is the run with Adam, bs 64, lr 0.0001, weight decay 0.001, where I got training loss 148150.33 and test loss 439392.578125
3 - I suspect that since you used mean value of the loss for the hessian calculation you get different range of numbers.

Please let me know if this helps.

**Comment by schen572** [comment]
Thank you for the reply!
Yes, I did mean reduction, because the training loss is too large to get to the criterion loss < 0.1 if no mean reduction is done. In your second point, did you mean that the training loss will eventually reduce from 148150.33 to below 0.1 (the criterion) after thousands of epochs? Under the scaling, doesn't 0.1 loss correspond to nearly perfect (100%) accuracy?

Also, I am wondering if the conclusion in the paper still holds even if bad generalization happens?

**Comment by link-er** [comment]
The mean reduction is there when you are training - and with mean loss it should go down to 0.1. But when you compute generalization gap and flatness measures you use loss without reduction.

Bad generalization should not affect the results, but underfitting will collapse generalization gap into almost zero (since both training and test loss will be equally high) and there generalization measures would not be helpful.

**Comment by schen572** [comment]
Then how is the generalization gap calculated in Figure 5? Is it the difference between mean train and test loss ? If so, shouldn't it be less than 1? 

Actually my screenshot shows that underfitting will not happen if no data augmentation is done. There is a substantial gen gap.

**Comment by link-er** [comment]
For the figure it is computed without reduction, please see the code here https://github.com/kampmichael/RelativeFlatnessAndGeneralization/blob/main/CorrelationFlatnessGeneralization/measure_comparison.py

Yes, indeed, maybe augmentation is needed for larger models. In any case, you can try to use this way of training as well - but I would not suggest comparing in one plot models trained on augmeneted and not augmented datasets since their loss surfaces should be different.

**Comment by schen572** [comment]
Sounds good, I don't have the capacity right now to test whether the loss will converge below 0.1, but I can try. It's helpful to know approximately how many epochs are required tho. Also, the plotting code is not working, which is why I was asking about the details of the plot (assemble_compare_measures is not defined in this [line](https://github.com/kampmichael/RelativeFlatnessAndGeneralization/blob/0baa1f0c87db2860e1a4d8f675ff0347a8872b3f/CorrelationFlatnessGeneralization/plot_measuresCompare.py#L15C103-L15C129)). 

Thanks for the help! You can close this issue whenever you see appropriate.

**Comment by link-er** [comment]
Oh, I see the missing file now! Sorry for that. Just added it, please let me know if it works now.

The convergence was achieved after approximately 2000 epochs if I remember correctly.
